src/coreclr/src/vm/dllimportcallback.cpp

```diff
@@ -626,32 +627,93 @@ VOID UMThunkMarshInfo::SetUpForUnmanagedCallersOnly()
 
     BYTE* pData = NULL;
     LONG cData = 0;
-    CorPinvokeMap callConv = (CorPinvokeMap)0;
 
+    bool nativeCallableInternalData = false;
     HRESULT hr = pMD->GetCustomAttribute(WellKnownAttribute::UnmanagedCallersOnly, (const VOID **)(&pData), (ULONG *)&cData);
     if (hr == S_FALSE)
+    {
         hr = pMD->GetCustomAttribute(WellKnownAttribute::NativeCallableInternal, (const VOID **)(&pData), (ULONG *)&cData);
+        nativeCallableInternalData = SUCCEEDED(hr);
+    }
 
     IfFailThrow(hr);
 
     _ASSERTE(cData > 0);
 
     CustomAttributeParser ca(pData, cData);
+
     // UnmanagedCallersOnly has two optional named arguments CallingConvention and EntryPoint.
     CaNamedArg namedArgs[2];
-    CaTypeCtor caType(SERIALIZATION_TYPE_STRING);
-    // First, the void constructor.
-    IfFailThrow(ParseKnownCaArgs(ca, NULL, 0));
-
-    // Now the optional named properties
-    namedArgs[0].InitI4FieldEnum("CallingConvention", "System.Runtime.InteropServices.CallingConvention", (ULONG)callConv);
-    namedArgs[1].Init("EntryPoint", SERIALIZATION_TYPE_STRING, caType);
-    IfFailThrow(ParseKnownCaNamedArgs(ca, namedArgs, lengthof(namedArgs)));
-
-    callConv = (CorPinvokeMap)(namedArgs[0].val.u4 << 8);
-    // Let UMThunkMarshalInfo choose the default if calling convension not definied.
-    if (namedArgs[0].val.type.tag != SERIALIZATION_TYPE_UNDEFINED)
-        m_callConv = (UINT16)callConv;
+
+    // For the UnmanagedCallersOnly scenario.
+    CaType caCallConvs;
+
+    // Define the optional named properties
+    if (nativeCallableInternalData)
+    {
+        namedArgs[0].InitI4FieldEnum("CallingConvention", "System.Runtime.InteropServices.CallingConvention", (ULONG)(CorPinvokeMap)0);
+    }
+    else
+    {
+        caCallConvs.Init(SERIALIZATION_TYPE_SZARRAY, SERIALIZATION_TYPE_TYPE, SERIALIZATION_TYPE_UNDEFINED, NULL, 0);
+        namedArgs[0].Init("CallConvs", SERIALIZATION_TYPE_SZARRAY, caCallConvs);
+    }
+
+    CaTypeCtor caEntryPoint(SERIALIZATION_TYPE_STRING);
```

Comment:
We are not using the EntryPoint for anything. Is it needed?

Reply:
Not at all. I was keeping the same reading logic. Let me remove it.

Reply:
It appears that the custom attribute reader API requires an entry if it is ever present. Looks like we need to keep it.